Add `apiclient network configure` subcommand

ytsssun opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes https://github.com/bottlerocket-os/bottlerocket-core-kit/issues/713

**Description of changes:**
Adds `apiclient network configure <URI>` command to configure network settings at runtime. The implementation includes:

  - API Server: `New /network/configure` POST endpoint that validates and writes net.toml content to `/.bottlerocket/net.toml`
  - API Client: New `network configure` subcommand supporting `file://` and `base64:` URI schemes
  - Input validation: UTF-8 validation, base64 decoding, and file access with proper error handling
  - Documentation: Updated README template and module documentation

**Testing done:**

**1. Tested the change via `bootstrap-commands` using `base64` scheme**
Started `metal-dev` with command
```bash
./tools/start-local-vm --variant metal-dev --arch $(uname -m) --inject-file net.toml --inject-file user-data.toml
```

My initial `net.toml` only has configurations for a single interface
```
version = 3

[enp0s16]
dhcp4 = true
primary = true
```

My `user-data.toml` uses `bootstrap-commands`
```
[settings.bootstrap-commands.setup-network-inline]
commands = [
  ["apiclient", "network", "configure", "base64:dmVyc2lvbiA9IDMKCltlbnAwczE2XQpkaGNwNCA9IHRydWUKZGhjcDYgPSBmYWxzZQpwcmltYXJ5ID0gdHJ1ZQoKW2VucDBzMTddCmRoY3A0ID0gdHJ1ZQpkaGNwNiA9IGZhbHNlCg=="],
  ["apiclient", "reboot"]
]
mode = "once"
essential = true
```
Where the base64 encoded net.toml is 
```bash
echo "dmVyc2lvbiA9IDMKCltlbnAwczE2XQpkaGNwNCA9IHRydWUKZGhjcDYgPSBmYWxzZQpwcmltYXJ5ID0gdHJ1ZQoKW2VucDBzMTddCmRoY3A0ID0gdHJ1ZQpkaGNwNiA9IGZhbHNlCg==" | 
base64 -d

version = 3

[enp0s16]
dhcp4 = true
dhcp6 = false
primary = true

[enp0s17]
dhcp4 = true
dhcp6 = false
```

Confirmed that the node automatically rebooted and the net.toml was written to the desired path /.bottlerocket/net.toml
```bash
bash-5.2# cat /.bottlerocket/net.toml 
version = 3

[enp0s16]
dhcp4 = true
dhcp6 = false
primary = true

[enp0s17]
dhcp4 = true
dhcp6 = false
```

And both interfaces were brought up
```bash
ip link show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: enp0s16: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
    link/ether 52:54:00:12:34:56 brd ff:ff:ff:ff:ff:ff
3: enp0s17: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
    link/ether 52:54:00:12:34:57 brd ff:ff:ff:ff:ff:ff
4: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN mode DEFAULT group default 
    link/ether 02:42:f1:1e:86:ce brd ff:ff:ff:ff:ff:ff
bash-5.2# cat /etc/systemd/network/10-enp0s16.network
```

**2. Tested the change via `bootstrap-container` using `file://` scheme**
Used similar setup as 1. with `metal-dev`. Only difference is the `user-data.toml`
```toml
[settings.bootstrap-containers.setup-network-file]
source = "public.ecr.aws/bottlerocket/bottlerocket-bootstrap:v0.2.8"
mode = "once"
user-data = "IyEvdXNyL2Jpbi9lbnYgYmFzaApzZXQgLWV1CgojIENoZWNrIHJlYm9vdCBtYXJrZXIgZmlyc3QgLSBza2lwIGlmIGFscmVhZHkgcHJvY2Vzc2VkCklOU1RBTkNFX1JFQk9PVEVEPS8uYm90dGxlcm9ja2V0L2Jvb3RzdHJhcC1jb250YWluZXJzL2N1cnJlbnQvcmVib290ZWQKaWYgWyAtZiAiJElOU1RBTkNFX1JFQk9PVEVEIiBdOyB0aGVuCiAgICBlY2hvICJDb25maWd1cmF0aW9uIGFscmVhZHkgcHJvY2Vzc2VkIgogICAgZXhpdCAwCmZpCgplY2hvICJTZXR0aW5nIHVwIG5ldHdvcmsgY29uZmlndXJhdGlvbiB2aWEgZmlsZTovLyBzY2hlbWUuLi4iCgojIENyZWF0ZSB0aGUgZXhhY3QgbmV0d29yayBjb25maWd1cmF0aW9uCmNhdCA+IC90bXAvbmV0d29yay1jb25maWcudG9tbCA8PCAnSU5ORVJFT0YnCnZlcnNpb24gPSAzCgpbZW5wMHMxNl0KZGhjcDQgPSB0cnVlCmRoY3A2ID0gZmFsc2UKcHJpbWFyeSA9IHRydWUKCltlbnAwczE3XQpkaGNwNCA9IHRydWUKZGhjcDYgPSBmYWxzZQpJTk5FUkVPRgoKZWNobyAiTmV0d29yayBjb25maWd1cmF0aW9uIGNyZWF0ZWQ6IgpjYXQgL3RtcC9uZXR3b3JrLWNvbmZpZy50b21sCgplY2hvICJBcHBseWluZyBuZXR3b3JrIGNvbmZpZ3VyYXRpb24gdXNpbmcgZmlsZTovLyBzY2hlbWUuLi4iCmFwaWNsaWVudCBuZXR3b3JrIGNvbmZpZ3VyZSBmaWxlOi8vL3RtcC9uZXR3b3JrLWNvbmZpZy50b21sCgplY2hvICJOZXR3b3JrIGNvbmZpZ3VyYXRpb24gYXBwbGllZCBzdWNjZXNzZnVsbHkiCgojIE1hcmsgYXMgcHJvY2Vzc2VkIHRvIHByZXZlbnQgcmVib290IGxvb3AKdG91Y2ggIiRJTlNUQU5DRV9SRUJPT1RFRCIKCmVjaG8gIlJlYm9vdGluZyB0byBhcHBseSBuZXR3b3JrIGNoYW5nZXMuLi4iCmFwaWNsaWVudCByZWJvb3QK"
```

Where the encoded user-data here is:
```bash
#!/usr/bin/env bash
set -eu

# Check reboot marker first - skip if already processed
INSTANCE_REBOOTED=/.bottlerocket/bootstrap-containers/current/rebooted
if [ -f "$INSTANCE_REBOOTED" ]; then
    echo "Configuration already processed"
    exit 0
fi

echo "Setting up network configuration via file:// scheme..."

# Create the exact network configuration
cat > /tmp/network-config.toml << 'INNEREOF'
version = 3

[enp0s16]
dhcp4 = true
dhcp6 = false
primary = true

[enp0s17]
dhcp4 = true
dhcp6 = false
INNEREOF

echo "Network configuration created:"
cat /tmp/network-config.toml

echo "Applying network configuration using file:// scheme..."
apiclient network configure file:///tmp/network-config.toml

echo "Network configuration applied successfully"

# Mark as processed to prevent reboot loop
touch "$INSTANCE_REBOOTED"

echo "Rebooting to apply network changes..."
apiclient reboot
```

With this approach a network-config.toml file was created within the container at `/tmp/network-config.toml` and we then used `apiclient network configure file:///tmp/network-config.toml` to configure the net.toml.

Note that for bootstrap-containers, a **marker file** is needed to prevent reboot loop. The reboot happens during the execution of the bootstrap script so the subsequent process that turns the `once` mode to `off` will never happen. 

`systemd-inhibit` [command](https://github.com/bottlerocket-os/bottlerocket-core-kit/blob/develop/packages/os/bootstrap-commands.service#L12) we used for bootstrap-commands is something we can look into which could be a way to avoid the usage of marker file. 

@KCSesh - Contribution: 

**1. Re-Rean the `bootstrap-commands` using `base64` scheme**

**2. Re-ran: Tested the change via `bootstrap-container` using `file://` scheme**
Note: Running this test multiple times required: `--force-extract`, to clean PRIVATE partition marker file

**3. Validated STDIN**

```
apiclient network configure <<EOF
version = 2
[eth0]
dhcp4 = true
primary = true
EOF
```

**4. Tested invalid input**
<details>

Invalid Version:
```
bash-5.2# journalctl -u bootstrap-commands
Nov 11 23:35:45 localhost systemd[1]: Starting Bootstrap Commands...
Nov 11 23:35:46 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1351]: 23:35:46 [INFO] Processing bootstrap command 'setup-network-invalid-version' ...
Nov 11 23:35:46 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1351]: 23:35:46 [WARN] Bootstrap command failed to execute 'apiclient' failed - stderr: Failed to configure network: Failed to POST network configuration to '/actions/network/configure': Status 400 when POSTing /actions/network/configure: Invalid network configuration: Failed to parse network config from stdin: Invalid network configuration: Unknown network config version: 99
Nov 11 23:35:46 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1351]: Bootstrap command 'setup-network-invalid-version' failed.
Nov 11 23:35:46 ip-10-0-2-15.us-west-2.compute.internal systemd-inhibit[1345]: /usr/bin/bootstrap-commands failed with exit status 1.
Nov 11 23:35:46 ip-10-0-2-15.us-west-2.compute.internal systemd[1]: bootstrap-commands.service: Main process exited, code=exited, status=1/FAILURE
Nov 11 23:35:46 ip-10-0-2-15.us-west-2.compute.internal systemd[1]: bootstrap-commands.service: Failed with result 'exit-code'.
Nov 11 23:35:46 ip-10-0-2-15.us-west-2.compute.internal systemd[1]: Failed to start Bootstrap Commands.

bash-5.2# journalctl -u apiserver
...
Nov 11 23:35:46 ip-10-0-2-15.us-west-2.compute.internal apiserver[1251]: 23:35:46 [INFO] Invoking netdog commit to validate and write network configuration
Nov 11 23:35:46 ip-10-0-2-15.us-west-2.compute.internal apiserver[1251]: 23:35:46 [ERROR] Netdog commit failed: Network configuration validation failed: Failed to parse network config from stdin: Invalid network configuration: Unknown network config version: 99

```

Unknown field: 

```
bash-5.2# journalctl -u bootstrap-commands
Nov 11 23:37:54 localhost systemd[1]: Starting Bootstrap Commands...
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1336]: 23:37:54 [INFO] Processing bootstrap command 'setup-network-unknown-field' ...
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1336]: 23:37:54 [WARN] Bootstrap command failed to execute 'apiclient' failed - stderr: Failed to configure network: Failed to POST network configuration to '/actions/network/configure': Status 400 when POSTing /actions/network/configure: Invalid network configuration: Failed to parse network config from stdin: Failed to parse network config: TOML parse error at line 2, column 7
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1336]:   |
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1336]: 2 | hello=hi
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1336]:   |       ^
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1336]: invalid string
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1336]: expected `"`, `'`
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal bootstrap-commands[1336]: Bootstrap command 'setup-network-unknown-field' failed.
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal systemd-inhibit[1330]: /usr/bin/bootstrap-commands failed with exit status 1.
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal systemd[1]: bootstrap-commands.service: Main process exited, code=exited, status=1/FAILURE
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal systemd[1]: bootstrap-commands.service: Failed with result 'exit-code'.
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal systemd[1]: Failed to start Bootstrap Commands.

bash-5.2# journalctl -u apiserver
...
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal apiserver[1238]: 23:37:54 [INFO] Invoking netdog commit to validate and write network configuration
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal apiserver[1238]: 23:37:54 [ERROR] Netdog commit failed: Network configuration validation failed: Failed to parse network config from stdin: Failed to parse network config: TOML parse error at line 2, column 7
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal apiserver[1238]:   |
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal apiserver[1238]: 2 | hello=hi
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal apiserver[1238]:   |       ^
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal apiserver[1238]: invalid string
Nov 11 23:37:54 ip-10-0-2-15.us-west-2.compute.internal apiserver[1238]: expected `"`, `'`
bash-5.2#

```

Invalid stdin:

```
# Starting point: 
bash-5.2# cat /.bottlerocket/net.toml
version = 3

[enp0s16]
dhcp4 = true
dhcp6 = false
primary = true

[enp0s17]
dhcp4 = true
dhcp6 = false

# Apply incorrect net.toml:
bash-5.2# apiclient network configure <<EOF
version = 3

[enp0s17]
dhcp5 = true
dhcp6 = false
EOF

[  191.895006] apiserver[1238]: 16:06:56 [ERROR] netdog commit failed: network configuration validation failed: Failed to parse network config from stdin: Failed to parse network config: data did not match any variant of untagged enum NetworkDeviceV1
Failed to configure network: Failed to POST network configuration to '/actions/network/configure': Status 400 when POSTing /actions/network/configure: Invalid network configuration: Failed to parse network config from stdin: Failed to parse network config: data did not match any variant of untagged enum NetworkDeviceV1

# No change to file: 
bash-5.2# cat /.bottlerocket/net.toml
version = 3

[enp0s16]
dhcp4 = true
dhcp6 = false
primary = true

[enp0s17]
dhcp4 = true
dhcp6 = false
bash-5.2#
```

</details>



**5. Verified new tmp approach, with admin-container and inotify on metal:**
```
[root@admin]# apiclient network configure <<EOF
> version = 3
> [enp0s16]
> dhcp4 = true
> dhcp6 = false
> primary = true
> EOF
IN_CREATE: /.bottlerocket/rootfs/.bottlerocket/.tmpZkeAI5
IN_OPEN: /.bottlerocket/rootfs/.bottlerocket/.tmpZkeAI5
IN_MODIFY: /.bottlerocket/rootfs/.bottlerocket/.tmpZkeAI5
IN_OPEN: /.bottlerocket/rootfs/.bottlerocket/.tmpZkeAI5
IN_MODIFY: /.bottlerocket/rootfs/.bottlerocket/.tmpZkeAI5
IN_CLOSE_WRITE: /.bottlerocket/rootfs/.bottlerocket/.tmpZkeAI5
IN_MOVED_FROM: /.bottlerocket/rootfs/.bottlerocket/.tmpZkeAI5
IN_MOVED_TO: /.bottlerocket/rootfs/.bottlerocket/net.toml
IN_CLOSE_WRITE: /.bottlerocket/rootfs/.bottlerocket/net.toml
```
**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,221 @@
+use tokio::io::AsyncReadExt;
+
+/// Configures network settings from the given input source (URI or stdin), writing the content
+/// directly to /.bottlerocket/net.toml via the API server.
```
Since this is the client, it doesn't care (and know) where the configuration file is written to.

```suggestion
/// Configures network settings from the given input source (URI or stdin)
```

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,221 @@
+where
+    P: AsRef<Path>,
+{
+    // Get the network configuration content from the input source
```
Comments are appreciated, but please review the comments and make sure they explain "why" you did what you did instead of the "what" and "how".

ytsssun replied:
That's fair, the code itself is pretty self-explanatory. Dropping this comment.

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,221 @@
+        tokio::io::stdin()
+            .read_to_string(&mut output)
+            .context(error::StdinReadSnafu)
+            .await?;
```
Is there any benefit of using futures here? There aren't any other tasks executed in the background and this call is blocking right away at this exact moment.

ytsssun replied:
Updated, thanks!

cbgbt replied:
It somewhat makes sense because `apiclient` is a lib crate and could be used in another asynchronous application in a concurrent fashion.

But I think exposing reading from stdin in the pub `network::configure` function is a bit odd -- it would be a very strange way to call the function from a consuming application.

Probably the public API should allow configuration with a string and the "stdin or file" element of `apiclient` should be kept private and only invoked via `main()`.

To be more explicit with my feedback, I think `network::configure()` should remain `pub` but should probably not call `get_content()` itself, and instead assume that the caller will do so.

ytsssun replied:
> It somewhat makes sense because apiclient is a lib crate and could be used in another asynchronous application in a concurrent fashion.

I was about to make the change to use synchronous call, but this actually makes sense to me. 

The `stdin` was not expected, I somehow had an old commit here.. 

But the idea to keep `configure()` and `get_content()` separate makes sense to me so that the caller will make two calls. I will keep the functions async in that case. And just update `configure()` to not call `get_content` instead.

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,221 @@
+        })?;
+        tokio::fs::read_to_string(path)
+            .context(error::FileReadSnafu { input_source })
+            .await
```
Same as elsewhere, if this is blocking right away, why making it an `async` call?

ytsssun replied:
Resolving as per @cbgbt if using this function via a crate, it could be used in async fashion, in which we should not block the runtime.

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,221 @@
+            source: std::io::Error,
+        },
+
+        #[snafu(display("Given invalid file URI '{}'", input_source))]
```
```suggestion
#[snafu(display("Invalid file URI '{}'", input_source))]
```

arnaldo2792 commented on sources/api/apiserver/src/server/mod.rs:
```diff
@@ -651,6 +652,30 @@ async fn reboot() -> Result<HttpResponse> {
+    if !config_dir.exists() {
+        std::fs::create_dir_all(config_dir)
+            .context(error::NetworkConfigDirSnafu { path: config_dir })?;
+    }
```
This program shouldn't be creating this directory, as it is expected to be managed by the system. Even if the program tries to create the directory, it will fail as the rootfs is read-only.

ytsssun replied:
Yeah I somehow had an old commit here.. You are right the mount point `/.bottlerocket` is managed by the mount unit we added for PRIVATE partition. `apiserver` should never bother creating the directory.

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,221 @@
+        response.text().await.context(error::ReqwestSnafu {
+            uri: input_source,
+            method: "GET",
+        })
```
You may want to have a different error for this, as it is passed the point of a failed `GET` request and it is at the point of parsing the text that was returned by the request.

ytsssun replied:
We should explicitly forbid this case as other schemes are not supported.

arnaldo2792 commented on sources/api/apiclient/src/main.rs:
```diff
@@ -225,6 +241,13 @@ fn usage() -> ! {
 
+        network configure options:
+            [ URI ]                    The URI to a network configuration file (TOML format) that you
+                                       want to apply. Supports file:// and base64: URI schemes.
```
According to the implementation, it also supports `http` and `https`.

ytsssun replied:
Ahh, great catch! `http` and `https` should not be in the scope. I have added explicit deny for this case. Right now we should only support file and base64. 

We will expand the support for broader input sources later, which I will follow up with a GH issue.

cbgbt commented on sources/api/apiserver/src/server/mod.rs:
```diff
@@ -651,6 +652,30 @@ async fn reboot() -> Result<HttpResponse> {
 }
 
+/// Configures network settings by writing content to /.bottlerocket/net.toml
+async fn configure_network(body: web::Bytes) -> Result<HttpResponse> {
```
One thing made clear by the client docs but not this source code is that we are writing the network configuration but that this API route expects for the caller to reboot for the changes to take effect.

ytsssun replied:
Fair ask, Will update the comment to mention that a reboot is expected to allow it to take effect.

ytsssun commented:
Force pushed to address comments from @cbgbt and @arnaldo2792

arnaldo2792 commented on sources/api/apiclient/README.md:
```diff
@@ -75,6 +75,53 @@ You can use JSON form to set it:
+
+#### Configuration format
+
+The net.toml file uses the same format that netdog supports. Here's an example:
```
nit: for consistency `net.toml` and not net.toml

arnaldo2792 commented on sources/api/apiclient/README.tpl:
```diff
@@ -75,6 +75,53 @@ You can use JSON form to set it:
+This allows you to configure network settings by providing a net.toml file which netdog uses for network configuration.
+The configuration is written to `/.bottlerocket/net.toml` and will be validated and applied at the next boot by netdog.
+
+The network configure command accepts input from different sources using URI schemes:
```
```suggestion
The `network configure` command accepts input from different sources using URI schemes:
```

cbgbt commented on sources/api/apiclient/src/main.rs:
```diff
@@ -1486,4 +1568,91 @@ mod tests {
+            _ => panic!(
+                "Expected Network::Configure subcommand: {expected_subcommand:?}, got: {subcommand:?}"
+            ),
+        }
```
The macro above is pretty effortful in making it easy to test new subcommands, but then we have to maintain this match tree.

Can we just `#[derive(PartialEq)]` on `Subcommand`?

ytsssun replied:
Done. I had to add `#derive(PartialEq)` to all the relevant structs referred by `Subcommand`.

ytsssun replied:
LMK what you think, I think that might not be worth it?

I now feel I should just reduce the test scope down to `test_network_configure_parsing`? I don't feel so good about adding `#derive(PartialEq)` to all the relevant structs, especially this is just for unit test.

cbgbt replied:
I think adding `#[derive(PartialEq)]` for relevant structs is a good idea, especially for tests.

It's pretty common to derive `Debug`, `Clone`, `Copy` (if applicable), `PartialEq`, and `Eq` if you can.

KCSesh replied:
Fixed in the latest push!

cbgbt commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,258 @@
+///
+/// Supports file:// and base64: URI schemes. Use this function to get content
+/// before calling `configure()`.
+pub async fn get_content<S>(input_source: S) -> Result<String>
```
minor nit: This should probably just be `pub(crate)`

ytsssun replied:
I cannot make it `pub(crate)` as the compiler sees the binary crate (`main.rs`) and lib crate (`lib.rs`) separate crates. Making this pub(crate) would make it invisible from `main.rs`. I think we should still keep this function within network.rs, in which case we have to keep get_content() as public function...

Alternative is to move get_content() to main.rs, but the functionality boundary is less clear get_content() was used to fetch the content. 

We have similar logic in `apply` where we use `get` as a private function. Another alternative is to revert to the approach where we call get_content() within configure() so that it remains private. 

My preference in order is:
1. Keep it `pub`.
2. Make it `pub(crate)` and move it to `main.rs`
3. Move back to the state where we make it private and call it within `configure()`

cbgbt replied:
I think keeping it `pub` is fine.

cbgbt commented on sources/api/apiclient/README.md:
```diff
@@ -75,6 +75,53 @@ You can use JSON form to set it:
+### Network configure
+
+This allows you to configure network settings by providing a net.toml file which netdog uses for network configuration.
+The configuration is written to `/.bottlerocket/net.toml` and will be validated and applied at the next boot by netdog.
```
Similar to what Arnaldo mentioned, I might drop references to `net.toml` from the client configuration, and just mention that it will be applied at next boot. Protects us from stale documentation if the underlying implementation changes later.

ytsssun replied:
Updated. I think it makes sense to not mention the exact path it will be written to. But mentioning `net.toml` and its format is probably needed. See the section below (#### Configuration format).

The user needs to know how to compose the `net.toml` as the source.

cbgbt reviewed (approved):
Nice work! just a few suggestions.

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,258 @@
+//! This module provides functionalities to configure network settings.
+//! It accepts the same net.toml format supported by netdog, but allows sourcing it from
+//! file:// and base64: URIs.
```
```suggestion
//! This module implements network configuration API calls.
//! It supports sourcing `net.toml` configuration files from the filesystem or base64 encoded strings.
```

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,258 @@
+use url::Url;
+
+/// Configures network settings by sending the provided content to the API server.
+///
+/// The configuration will be applied at the next boot - a reboot is required for changes to take effect.
```
nit: I wouldn't put this here as this is relevant to the user (which won't read this), and rather should be in the website docs or the main `apiclient` documentation.

cbgbt replied:
I personally found it helpful, without this comment I spent time looking around after seeing the apiserver implementation to try to understand if we were missing the code to actually apply the config after saving it.

But I've also been on the unpopular side of this same tension many times, so I'm curious to see how others think about it.

ytsssun replied:
I don't have strong opinion here and I could see the tradeoffs. But this change is made to address comments here - https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/714#discussion_r2505679409

So I would like to keep it as is now to avoid extra churns.

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,258 @@
+/// Retrieves the network configuration content from the given source URI.
+///
+/// Supports file:// and base64: URI schemes. Use this function to get content
+/// before calling `configure()`.
```
```suggestion
/// Supports file:// and base64: URI schemes
```

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,258 @@
+        assert!(result.is_err());
+        let error_msg = result.unwrap_err().to_string();
+        assert!(error_msg.contains("Unsupported URI scheme 'http'"));
+        assert!(error_msg.contains("Only file:// and base64: schemes are supported"));
```
neat!

arnaldo2792 commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    #[tokio::test]
+    async fn test_https_uri_rejected() {
```
Can you use your awesome `test_case` here and have the one test for the different use cases?

ytsssun replied:
Done.

ytsssun committed:
apiclient: add network configure subcommand

Implements 'apiclient network configure <input-source>' to enable runtime
network configuration for Bottlerocket. stdin is the default input, but also
supports file://, base64: input sources with client-side URI processing.
Content is sent to the API server for further processing.

Signed-off-by: Yutong Sun <yutongsu@amazon.com>
Signed-off-by: Kyle Sessions <kssessio@amazon.com>

ytsssun commented:
Force pushed to address comments from @cbgbt and @arnaldo2792 :
1. Remove comments that is redundant where the code is self-explanatory.
2. Documents wording update.

bcressey commented on sources/api/apiserver/src/server/mod.rs:
```diff
@@ -687,6 +688,31 @@ async fn reboot() -> Result<HttpResponse> {
+
+    // Write the configuration content to the file
+    std::fs::write(&config_file, content)
+        .context(error::NetworkConfigWriteSnafu { path: &config_file })?;
```
Given the high potential for bricking the system, I'd prefer we have a `netdog validate` command that ensures it's a valid, deserializable configuration.

bcressey commented on sources/api/apiclient/src/main.rs:
```diff
@@ -564,6 +587,46 @@ fn parse_lockdown_args(args: Vec<String>) -> Subcommand {
+
+    let input_source = input_source.unwrap_or_else(|| {
+        usage_msg("apiclient network configure requires an input source URI (file:// or base64:)")
+    });
```
It'd be nice to default to stdin here, as `apiclient apply` does, so we can run:
```
apiclient network configure <<EOF
version = 2
[eth0]
dhcp4 = true
primary = true
EOF
```

KCSesh committed:
netdog: add commit subcommand

Adds 'netdog commit' subcommand that validates and writes network
configuration from stdin to /.bottlerocket/net.toml.

Signed-off-by: Kyle Sessions <kssessio@amazon.com>

cbgbt commented on sources/netdog/src/cli/commit.rs:
```diff
@@ -0,0 +1,40 @@
+pub struct CommitArgs {}
+
+pub fn run(_args: CommitArgs) -> Result<()> {
+    // Read content from stdin
```
Mind purging these "what" comments?

KCSesh replied:
Removed the what commets from commit.rs

cbgbt commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,262 @@
+
+/// Internal function that accepts a custom reader
+async fn get_content_with_reader<S, R>(input_source: Option<S>, mut reader: R) -> Result<String>
+where
```
Mind updating this docstring to be more descriptive? Ideally should declare the function's purpose in isolation.

KCSesh replied:
Updated, and it was also refactored a bit on the previous push!

cbgbt commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,262 @@
+where
+    S: Into<String>,
+{
+    get_content_with_reader(input_source, tokio::io::stdin()).await
```
It feels somewhat like `get_content` should be deciding whether or not to use the `input_source` or `stdin` rather than passing both to the delegate function and allowing it to decide.

KCSesh replied:
Previous push, should handle this

KCSesh commented:
^ This is a decent diff URL:
https://github.com/bottlerocket-os/bottlerocket-core-kit/compare/d5a84546284dd157011259800a3debbeac32ecde..4facba9d46b4f48afc45a1d75471fbda13139cf4

Above I added 2 minor changes:
* Add `netdog commit` which gets invoked by `apiserver` after `apiclient` does initial parse logic. Netdog commit now does the write to `/.bottlerocket/net.toml` 
* Add `stdin` support to `apiclient network configure`

I retested both valid and invalid net.toml files on `metal-dev`  (will update the description).

KCSesh commented:
^ The above push is just a rebase to get back on/near the tip of `develop`

KCSesh commented:
^ Fix comment typo, renamed 1 error, and created a const.

bcressey commented on sources/netdog/src/cli/commit.rs:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    std::fs::write(&config_file, content)
+        .context(error::NetConfigWriteSnafu { path: &config_file })?;
```
It'd be good to follow the pattern of writing to a temp file in the same directory, then rename it to the final file. Helps with power safety. See `persist_sysctl_config()` in `corndog` for an example.

KCSesh replied:
Nice!

bcressey commented on sources/api/apiserver/src/server/mod.rs:
```diff
@@ -687,6 +688,44 @@ async fn reboot() -> Result<HttpResponse> {
+        .stderr(std::process::Stdio::piped())
+        .spawn()
+        .and_then(|mut child| {
+            use std::io::Write;
```
```suggestion
```

KCSesh replied:
Thanks!

bcressey commented on sources/api/apiclient/src/network.rs:
```diff
@@ -0,0 +1,282 @@
+/// Retrieves the network configuration content from the given source URI.
+///
+/// Supports file:// and base64: URI schemes. If no input source is provided, reads from stdin.
+pub async fn get_content<S>(input_source: Option<S>) -> Result<String>
+where
+    S: Into<String>,
+{
+    get_content_with_reader(input_source, tokio::io::stdin()).await
+}
```
It might be more readable to have this fork into two functions, like `get_content_with_stdin` or `get_content_from_source`.

Then `get_content_from_source` could further split into `get_content_from_file`, `get_content_from_base64`.

I found it hard to quickly answer questions like "can I read base64-encoded data from a file or URI, or only from the command line?" - because of all the branching in `get_content_with_reader`.

(It'd also be great to support the base64: syntax for `apiclient apply`, for overall UI consistency.)

KCSesh replied:
Sounds good - made the change for now, and I'll add a tracker in apiclient apply to support base64

cbgbt reviewed (commented):
Apologies, I reviewed this yesterday but never hit submit.

Do you think you could fix the PartialEq comment left earlier?

KCSesh committed:
apiserver: add network configure endpoint with netdog integration

Adds POST /actions/network/configure endpoint to write network configuration
content directly to /.bottlerocket/net.toml. Includes error handling for
UTF-8 validation, directory creation, and file writing operations.

The endpoint integrates with netdog commit functionality to apply network
configuration changes.

Signed-off-by: Yutong Sun <yutongsu@amazon.com>
Signed-off-by: Kyle Sessions <kssessio@amazon.com>

KCSesh committed:
apiclient: derive PartialEq for subcommand types

Add PartialEq derivation to Subcommand and all related argument structs
to enable direct equality comparison in tests. Simplifies network configure
tests by replacing match statements with assert_eq comparisons.

Signed-off-by: Kyle Sessions <kssessio@amazon.com>

KCSesh commented:
@cbgbt 
>Do you think you could fix the PartialEq comment left earlier?

Sorry missed this - fixed in the latest push

Along with the other nits (update docstring and remove what comments from netdog commit)
